source/man/kirbysetup.1
'\" t
.\"     Title: Kirbysetup
.\"    Author: Uwe Gehring <adspectus@fastmail.com>
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 06/02/2020
.\"    Manual: kirbysetup User Manual
.\"    Source: kirbysetup
.\"  Language: English
.\"
.TH "KIRBYSETUP" "1" "06/02/2020" "kirbysetup" "kirbysetup User Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
kirbysetup \- Script to setup a new Kirby website
.SH "SYNOPSIS"
.HP \w'\fBkirbysetup\fR\ 'u
\fBkirbysetup\fR [\fB\-d\fR]
.HP \w'\fBkirbysetup\fR\ 'u
\fBkirbysetup\fR [\fB\-h\fR]
.SH "DESCRIPTION"
.PP
This manual page documents briefly the
\fBkirbysetup\fR
command, which is part of the
\fBkirbytools\fR(1)
package\&.
.PP
\fBkirbysetup\fR
is a script that will set up a new Kirby website\&. The installation will be done by means of the
\fBkirbyinstall\fR(1)
script\&. The selected package will be downloaded first if it is not already on the local machine by means of the
\fBkirbydownload\fR(1)
script\&.
.PP
All necessary information will be gathered interactively with the default values set by the
\fBkirbyconfigure\fR(1)
script, which will be called upon first invocation of any of the
kirbytools
scripts\&. The values will be saved in the file
$HOME/\&.kirbyrc\&.
.PP
Hence, if you have deliberately defined your default values therein, and do not want to deviate, you can just confirm all settings\&. The script will present you a summary of your settings and give you the opportunity to cancel the setup without changing anything on your system\&.
.PP
In addition to what can be done by means of the
\fBkirbydownload\fR
and
\fBkirbyinstall\fR
commands, the script gives you the option to either create your default Kirby panel admin user or enable the panel\&.
.PP
Furthermore, it will look into the
templates
subdirectory of your apache configuration files root folder (as configured by the
\fBkirbyconfigure\fR
command) and \- if it find appropriate named template files \- will create virtual host configuration files for the just created Kirby website\&.
.SH "OPTIONS"
.PP
The script follows the usual GNU command line syntax, options always starting with a dash (`\-\*(Aq)\&.
.PP
\fB\-h\fR
.RS 4
Show summary of options\&.
.RE
.PP
\fB\-d\fR
.RS 4
Show debugging information\&.
.RE
.SH "CREATING YOUR DEFAULT KIRBY PANEL ADMIN USER"
.PP
TODO
.SH "ENABLING THE PANEL"
.PP
TODO
.SH "CREATING VIRTUAL HOST CONFIGURATION FILES"
.PP
TODO
.SH "FILES"
.PP
/etc/kirbytools/kirbyrc
.RS 4
The system\-wide configuration file to control the behaviour of
kirbysetup\&.
.RE
.PP
${HOME}/\&.kirbyrc
.RS 4
The per\-user configuration file to control the behaviour of
kirbysetup\&. This file will be created by the
\fBkirbyconfigure\fR
script (see
\fBkirbyconfigure\fR(1)) if it does not exist\&.
.RE
.SH "BUGS"
.PP
Currently, only Kirby 3 starterkits and plainkits are supported\&.
.PP
The programs are currently optimized to work with the
apache2
httpd\&.
.PP
Some functionality requires root (or sudo as root) access\&.
.SH "SEE ALSO"
.PP
\fBkirbyconfigure\fR(1),
\fBkirbydeinstall\fR(1),
\fBkirbydownload\fR(1),
\fBkirbyinstall\fR(1),
\fBkirbytools\fR(1)
.SH "AUTHOR"
.PP
\fBUwe Gehring\fR <\&adspectus@fastmail\&.com\&>
.RS 4
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2020 Uwe Gehring
.br
.PP
This manual page was written for the Debian system (and may be used by others)\&.
.PP
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 2 or (at your option) any later version published by the Free Software Foundation\&.
.PP
On Debian systems, the complete text of the GNU General Public License can be found in
/usr/share/common\-licenses/GPL\&.
.sp


kirbytools_1.0/source/kirbySetup.sh
#!/bin/bash
#
# kirbySetup by Uwe Gehring <uwe@imap.cc>

## Read default variables
if [ -f /etc/kirbytools/kirbyrc ];then
  . /etc/kirbytools/kirbyrc
else
  echo "File /etc/kirbytools/kirbyrc not found!" && exit 2
fi

## Read default functions
if [ -f /etc/kirbytools/kirbyfunctions ];then
  . /etc/kirbytools/kirbyfunctions
else
  echo "File /etc/kirbytools/kirbyfunctions not found!" && exit 2
fi

## Initialize settings
initKirbySetup

## Usage
usage() {
  echo ""
  echo "  Usage:"
  echo ""
  echo "  $(basename $0) [-h] [-d]"
  echo ""
  echo "  -h: Show this help"
  echo "  -d: Show debug information"
  echo ""
  echo "  $(basename $0) will ask for all settings and can be exited before"
  echo "  doing anything in the system. Run with ${txtbld}-d${txtrst} to see all settings."
  echo ""
  exit 1
}

## Get commandline options
while getopts ":dh" opt;do
  case "${opt}" in
    d)  DEBUG=1;;
    h)  usage;;
    :)  errMsg "-$OPTARG requires an argument" && usage;;
    *)  errMsg "Invalid option $OPTARG" && usage;;
  esac
done
shift $((OPTIND -1))

## Ask all settings
askKit
askVersion
askVHost
askLink
askAccount
askPanel
askConf

KIRBYSELECTEDVHOSTDIR=$(getVHostDir $KIRBYSELECTEDVHOST)
KIRBYSELECTEDVHOSTLOGDIR=$(getLogDir $KIRBYSELECTEDVHOST)

## Show all variables (in debug mode) and all settings
[[ $DEBUG ]] && showVars
showSettings

## Allow to stop here without doing anything (default: continue)
read -n1 -p "${txtbld}If you are satisfied with above settings, continue? [${txtrst}${txtblue}Y${txtrst}${txtbld}|n] ${txtrst}"
[[ -z $REPLY ]] || echo ""
SEL=${REPLY:-y}
[[ "$SEL" == "y" || "$SEL" == "Y" ]] || exit 0

## Download Kit if not already here
echo -en "\nDownloading Kirby $KIRBYSELECTEDKIT-$KIRBYSELECTEDVERSION... "
KirbyDownload -k $KIRBYSELECTEDKIT -v $KIRBYSELECTEDVERSION -t $KIRBYDOWNLOADDIR > /dev/null 2>&1
[[ $? -eq 0 ]] && echo -e "${txtgreen}successful.${txtrst}" || echo -e "${txtred}failed.${txtrst}\n"

## Install Kit to Virtual Host Dir
echo -n "Installing Kirby $KIRBYSELECTEDKIT-$KIRBYSELECTEDVERSION... "
if [ "$KIRBYSELECTEDKIRBYLINK" == "Yes" ];then
## and replacing the 'kirby' folder with a symbolic link
  KirbyInstall -l -p $KIRBYSELECTEDKIT-$KIRBYSELECTEDVERSION -w $KIRBYSELECTEDVHOST > /dev/null 2>&1
else
## and NOT replacing the 'kirby' folder with a symbolic link
  KirbyInstall -p $KIRBYSELECTEDKIT-$KIRBYSELECTEDVERSION -w $KIRBYSELECTEDVHOST > /dev/null 2>&1
fi
[[ $? -eq 0 ]] && echo -e "${txtgreen}successful.${txtrst}" || echo -e "${txtred}failed.${txtrst}\n"

## Create admin account if selected and if script exist
if [ "$KIRBYSELECTEDCREATEACCOUNT" == "Yes" -a -e $KIRBYCREATEUSERSCRIPT ];then
  echo -n "Creating default account... "
  php $KIRBYCREATEUSERSCRIPT "$KIRBYSELECTEDVHOSTDIR" "$KIRBYADMINUSERMAIL" "$KIRBYADMINUSERNAME" "$KIRBYADMINUSERLANG" "$KIRBYADMINUSERPASS"
  [[ $? -eq 0 ]] && echo -e "${txtgreen}successful.${txtrst}" || echo -e "${txtred}failed.${txtrst}\n"
fi

## Enable the panel if selected
if [ "$KIRBYSELECTEDENABLEPANEL" == "Yes" ];then
  echo -n "Enabling the panel... "
  if [ "$KIRBYSELECTEDKIT" == "starterkit" ];then
    sed -i.bak -e "s/\[/\[ 'panel' => \[ 'install' => true \],/" $KIRBYSELECTEDVHOSTDIR/site/config/config.php
  else
    mkdir -p $KIRBYSELECTEDVHOSTDIR/site/config
    echo -e "<?php\n\nreturn [ 'panel' => [ 'install' => true ],'debug' => true ];\n" > $KIRBYSELECTEDVHOSTDIR/site/config/config.php
  fi
  [[ $? -eq 0 ]] && echo -e "${txtgreen}successful.${txtrst}" || echo -e "${txtred}failed.${txtrst}\n"
fi

## Copying templates as .conf files to sites-available dir in KIRBYAPACHECONFDIR
for tpl in ${_KIRBYTEMPLATES[@]};do
  TMPL="$KIRBYTEMPLATEDIR/$tpl.template"
  CONF="$KIRBYAPACHECONFDIR/sites-available/${tpl/vhost/$KIRBYSELECTEDVHOST}.conf"
  echo -n "Creating $(basename $CONF)... "

  [[ -f $CONF ]] && mv $CONF $CONF.bak

  echo -e "# Virtual Host ${tpl/vhost-/} Configuration created by $(basename $0) $KIRBYTOOLSVERSION\n# Name: $KIRBYSELECTEDVHOSTNAME\n# Date: $(date)\n# Desc: $KIRBYSELECTEDVHOSTDESC\n" > $CONF
  cat $TMPL >> $CONF

  sed -i -e "s/<VHOST>/$KIRBYSELECTEDVHOST/g" $CONF
  [[ $? -eq 0 ]] && echo -e "${txtgreen}successful.${txtrst}" || echo -e "${txtred}failed.${txtrst}\n"
done

if [ "$KIRBYSELECTEDCOMBINECONF" == "Yes" ];then
  if [ -f $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST-Redirection.conf -a -f $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST-SSL.conf ];then 
    echo -n "Merging redirection and ssl vhost configuration files into 1 file... "
    cat $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST-Redirection.conf $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST-SSL.conf > $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST.conf
    if [ -f $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST.conf ];then
      rm $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST-Redirection.conf $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST-SSL.conf
    fi
    [[ $? -eq 0 ]] && echo -e "${txtgreen}successful.${txtrst}" || echo -e "${txtred}failed.${txtrst}\n"
  fi
fi

if $(which vhostEnable);then
  for conf in $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST*.conf;do
    echo ""
    CONF=$(basename $conf .conf)
    read -n1 -p "${txtbld}Enable Virtual Host $CONF? [y|${txtrst}${txtblue}N${txtrst}${txtbld}] ${txtrst}"
    [[ -z $REPLY ]] || echo ""
    SEL=${REPLY:-n}
    if [[ "$SEL" == "y" || "$SEL" == "Y" ]];then
      echo -n "Enabling Virtual Host $CONF ..."
      vhostEnable $CONF > /dev/null
      CONFIGTEST=$(sudo apache2ctl configtest 2>&1 > /dev/null)
      ERR=$?
      if [ "$ERR" != "0" ];then
        echo -e "${txtred}failed.${txtrst}\n"
        errMsg $CONFIGTEST
      else
        sudo apache2ctl graceful
        echo -e "${txtgreen}successful.${txtrst}"
      fi
    fi
  done
else
  echo ""
  echo "To enable the virtual host, create a symbolic link in $KIRBYAPACHECONFDIR/sites-enabled/, i.e."
  echo "  ln -s $KIRBYAPACHECONFDIR/sites-available/$KIRBYSELECTEDVHOST.conf $KIRBYAPACHECONFDIR/sites-enabled/"
  echo "and restart apache with i.e."
  echo "  sudo apache2ctl graceful"
  echo ""
fi

exit 0
## code: language=bash